src/connectors/pagination/connectPagination.ts

```diff
@@ -0,0 +1,187 @@
+  nbHits: number;
+
+  /** The number of pages for the result set. */
+  nbPages: number;
+
+  /** The actual pages relevant to the current situation and padding. */
+  pages: number[];
+
+  /** true if the current page is also the first page. */
+  isFirstPage: boolean;
+
+  /** true if the current page is also the last page. */
+  isLastPage: boolean;
+};
+
+type ConnectPagination = Connector<
+  PaginationRendererOptions,
+  PaginationConnectorParams
+>;
+
+/**
+ * **Pagination** connector provides the logic to build a widget that will let the user
+ * choose the current page of the results.
+ *
+ * When using the pagination with Algolia, you should be aware that the engine won't provide you pages
+ * beyond the 1000th hits by default. You can find more information on the [Algolia documentation](https://www.algolia.com/doc/guides/searching/pagination/#pagination-limitations).
+ */
+const connectPagination: ConnectPagination = function connectPagination(
+  renderFn,
+  unmountFn = noop
+) {
+  checkRendering(renderFn, withUsage());
+
+  return widgetParams => {
+    const { totalPages, padding = 3 } =
+      widgetParams || ({} as typeof widgetParams);
+
+    const pager = new Paginator({
+      currentPage: 0,
+      total: 0,
+      padding,
+    });
+
+    type ConnectorState = {
+      refine(page: number): void;
+      createURL(state: SearchParameters): (page: number) => string;
+    };
+
+    const connectorState = {} as ConnectorState;
+
+    function getMaxPage({ nbPages }) {
+      return totalPages !== undefined ? Math.min(totalPages, nbPages) : nbPages;
+    }
+
+    return {
+      $$type: 'ais.pagination',
+
+      init(initOptions) {
+        const { instantSearchInstance } = initOptions;
+
+        renderFn(
+          {
+            ...this.getWidgetRenderState(initOptions),
+            instantSearchInstance,
+          },
+          true
+        );
+      },
+
+      render(renderOptions) {
+        const { instantSearchInstance } = renderOptions;
+
+        renderFn(
+          {
+            ...this.getWidgetRenderState(renderOptions),
+            instantSearchInstance,
+          },
+          false
+        );
+      },
+
+      dispose({ state }) {
+        unmountFn();
+
+        return state.setQueryParameter('page', undefined);
+      },
+
+      getWidgetUiState(uiState, { searchParameters }) {
+        const page = searchParameters.page || 0;
+
+        if (!page) {
+          return uiState;
+        }
+
+        return {
+          ...uiState,
+          page: page + 1,
+        };
+      },
+
+      getWidgetSearchParameters(searchParameters, { uiState }) {
+        const page = uiState.page ? uiState.page - 1 : 0;
+
+        return searchParameters.setQueryParameter('page', page);
+      },
+
+      getWidgetRenderState({ results, helper, createURL }) {
+        if (!connectorState.refine) {
+          connectorState.refine = page => {
```

Comment:
It would be nice to have such a syntax:

```ts
connectorState.refine ??= page => { /* ... */ }
```

Reply:
https://tc39.es/proposal-logical-assignment/#sec-assignment-operators

Reply:
I just tried and it requires `@babel/plugin-proposal-logical-assignment-operators` to be added. I can create a separate PR with this if you want.

Reply:
I think we can decide on that later